src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs

```diff
@@ -262,54 +262,82 @@ private Type0CommandMessageSection<BsonDocument> CreateType0Section(ConnectionDe
                 && _readPreference != null
                 && _readPreference != ReadPreference.Primary)
             {
                 var readPreferenceDocument = QueryHelper.CreateReadPreferenceDocument(_readPreference);
-                addIfNotAlreadyAdded("$readPreference", readPreferenceDocument);
+                AddIfNotAlreadyAdded("$readPreference", readPreferenceDocument);
             }
 
             if (_session.Id != null)
             {
-                addIfNotAlreadyAdded("lsid", _session.Id);
+                if (IsSessionUnacknowledged())
+                {
+                    if (_session.IsImplicit)
+                    {
+                        // do not set sessionId If session is implicit and write is unacknowledged
+                    }
+                    else
+                    {
+                        throw new NotSupportedException("Explicit session must not be used with unacknowledged writes.");
+                    }
+                }
+                else
+                {
+                    AddIfNotAlreadyAdded("lsid", _session.Id);
+                }
             }
 
             if (_session.ClusterTime != null)
             {
-                addIfNotAlreadyAdded("$clusterTime", _session.ClusterTime);
+                AddIfNotAlreadyAdded("$clusterTime", _session.ClusterTime);
             }
             Action<BsonWriterSettings> writerSettingsConfigurator = s => s.GuidRepresentation = GuidRepresentation.Unspecified;
 
             _session.AboutToSendCommand();
             if (_session.IsInTransaction)
             {
                 var transaction = _session.CurrentTransaction;
-                addIfNotAlreadyAdded("txnNumber", transaction.TransactionNumber);
+                AddIfNotAlreadyAdded("txnNumber", transaction.TransactionNumber);
                 if (transaction.State == CoreTransactionState.Starting)
                 {
-                    addIfNotAlreadyAdded("startTransaction", true);
+                    AddIfNotAlreadyAdded("startTransaction", true);
                     var readConcern = ReadConcernHelper.GetReadConcernForFirstCommandInTransaction(_session, connectionDescription);
                     if (readConcern != null)
                     {
-                        addIfNotAlreadyAdded("readConcern", readConcern);
+                        AddIfNotAlreadyAdded("readConcern", readConcern);
                     }
                 }
-                addIfNotAlreadyAdded("autocommit", false);
+                AddIfNotAlreadyAdded("autocommit", false);
             }
 
             var elementAppendingSerializer = new ElementAppendingSerializer<BsonDocument>(BsonDocumentSerializer.Instance, extraElements, writerSettingsConfigurator);
             return new Type0CommandMessageSection<BsonDocument>(_command, elementAppendingSerializer);
 
-            void addIfNotAlreadyAdded(string key, BsonValue value)
+            void AddIfNotAlreadyAdded(string key, BsonValue value)
```

Comment:
As far as I remember, we decided to use lowerCamelCase for local method names.

Reply:
In summary: I agree with this renaming. LGTM.